src/Lib.P2P/Routing/Dht1.cs

```diff
@@ -0,0 +1,380 @@
+
+        /// <summary>
+        ///   Peers that can provide some content.
+        /// </summary>
+        public ContentRouter ContentRouter;
+
+        /// <summary>
+        ///   The number of closer peers to return.
+        /// </summary>
+        /// <value>
+        ///   Defaults to 20.
+        /// </value>
+        public int CloserPeerCount { get; set; } = 20;
+
+        /// <summary>
+        ///   Raised when the DHT is stopped.
+        /// </summary>
+        /// <seealso cref="StopAsync"/>
+        public event EventHandler Stopped;
+
+        /// <inheritdoc />
+        public override string ToString() { return $"/{Name}/{Version}"; }
+
+        /// <inheritdoc />
+        public async Task ProcessMessageAsync(PeerConnection connection,
+            Stream stream,
+            CancellationToken cancel = default(CancellationToken))
+        {
+            while (true)
+            {
+                var request = await ProtoBufHelper.ReadMessageAsync<DhtMessage>(stream, cancel).ConfigureAwait(false);
+
+                log.Debug($"got {request.Type} from {connection.RemotePeer}");
+                var response = new DhtMessage
+                {
+                    Type = request.Type,
+                    ClusterLevelRaw = request.ClusterLevelRaw
+                };
+                switch (request.Type)
+                {
+                    case MessageType.Ping:
+                        response = ProcessPing(request, response);
+                        break;
+                    case MessageType.FindNode:
+                        response = ProcessFindNode(request, response);
+                        break;
+                    case MessageType.GetProviders:
+                        response = ProcessGetProviders(request, response);
+                        break;
+                    case MessageType.AddProvider:
+                        response = ProcessAddProvider(connection.RemotePeer, request, response);
+                        break;
+                    default:
+                        log.Debug($"unknown {request.Type} from {connection.RemotePeer}");
+
+                        // TODO: Should we close the stream?
+                        continue;
+                }
+
+                if (response != null)
+                {
+                    Serializer.SerializeWithLengthPrefix(stream, response, PrefixStyle.Base128);
+                    await stream.FlushAsync(cancel).ConfigureAwait(false);
+                }
+            }
+        }
+
+        /// <inheritdoc />
+        public Task StartAsync()
+        {
+            log.Debug("Starting");
+
+            RoutingTable = new RoutingTable(Swarm.LocalPeer);
+            ContentRouter = new ContentRouter();
+            Swarm.AddProtocol(this);
+            Swarm.PeerDiscovered += Swarm_PeerDiscovered;
+            Swarm.PeerRemoved += Swarm_PeerRemoved;
+            foreach (var peer in Swarm.KnownPeers) RoutingTable.Add(peer);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'foreach' block.](https://app.codacy.com/app/catalyst-network/Catalyst.Framework/pullRequest?prid=4612924)